### ISSUE TITLE
Crash when broadcasting WeatherSpec with too many locations

### ISSUE THREAD

**Issue by joserebelo** [breezy-weather/breezy-weather#1349]
### Steps to reproduce

1. Add multiple (> 5) locations to Breezy Weather
2. Enable WeatherSpec broadcast in settings

### Expected behavior

Neither sender nor receiver should crash

### Actual behavior

Over 3~5 locations makes the target app (Gadgetbridge) crash, depending on the number of daily/hourly forecasts:

```
08-10 13:52:20.097 W/BpBinder( 1396): Large outgoing transaction of 535260 bytes, interface descriptor , code 14
08-10 13:52:20.097 W/ActivityManager( 1396): pid 1396 system sent binder code 14 with flags 1 to frozen apps and got error -2147483646
08-10 13:52:20.097 E/JavaBinder( 1396): !!! FAILED BINDER TRANSACTION !!!  (parcel size = 535260)
08-10 13:52:20.097 W/BroadcastQueue( 1396): Failed to schedule BroadcastRecord{298ed9c nodomain.freeyourgadget.gadgetbridge.ACTION_GENERIC_WEATHER/u10} to BroadcastFilter{645b921 1010230/u10 ReceiverList{dc3fb88 29506 nodomain.freeyourgadget.gadgetbridge.nightly_nopebble/1010230/u10 remote:7f7bb2b}} via ProcessRecord{d1b0a5f 29506:nodomain.freeyourgadget.gadgetbridge.nightly_nopebble/u10a230}: android.os.TransactionTooLargeException: data parcel size 535260 bytes
```

https://codeberg.org/Freeyourgadget/Gadgetbridge/issues/3976

Over 8 locations makes Breezy Weather itself crash:

```
Caused by: java.lang.RuntimeException: android.os.TransactionTooLargeException: data parcel size 1144816 bytes
    at android.app.ContextImpl.sendBroadcast(ContextImpl.java:1290)
    at android.content.ContextWrapper.sendBroadcast(ContextWrapper.java:513)
    at android.content.ContextWrapper.sendBroadcast(ContextWrapper.java:513)
    at org.breezyweather.sources.RefreshHelper.broadcastDataIfNecessary(RefreshHelper.kt:886)
    at org.breezyweather.sources.RefreshHelper.broadcastDataIfNecessary$default(RefreshHelper.kt:851)
    at org.breezyweather.main.MainActivityViewModel$refreshBackgroundViews$1$1.invokeSuspend$lambda$0(MainActivityViewModel.kt:585)
    at org.breezyweather.main.MainActivityViewModel$refreshBackgroundViews$1$1.$r8$lambda$AXsl0jmtLEAPgOIVUNnttyhfxyM(Unknown Source:0)
    at org.breezyweather.main.MainActivityViewModel$refreshBackgroundViews$1$1$$ExternalSyntheticLambda0.run(D8$$SyntheticClass:0)
    at org.breezyweather.common.utils.helpers.AsyncHelper.delayRunOnIO$lambda$2(AsyncHelper.kt:56)
    at org.breezyweather.common.utils.helpers.AsyncHelper.$r8$lambda$5b3nePoL_LvCtLHM9YQ2MmyMu9Y(Unknown Source:0)
    at org.breezyweather.common.utils.helpers.AsyncHelper$$ExternalSyntheticLambda2.run(D8$$SyntheticClass:0)
    at io.reactivex.rxjava3.internal.operators.observable.ObservableDoOnEach$DoOnEachObserver.onComplete(ObservableDoOnEach.java:133)
    ... 10 more
Caused by: android.os.TransactionTooLargeException: data parcel size 1144816 bytes
    at android.os.BinderProxy.transactNative(Native Method)
    at android.os.BinderProxy.transact(BinderProxy.java:662)
    at android.app.IActivityManager$Stub$Proxy.broadcastIntentWithFeature(IActivityManager.java:6249)
    at android.app.ContextImpl.sendBroadcast(ContextImpl.java:1285)
    ... 21 more
```

### Weather source used

Open-Meteo

### Breezy Weather version

Self-built, e1bb8713e

### Android version

Android 14

### Device

Samsung Galaxy S23+

### Other details

The quickest fix is probably to limit the number of locations that gets broadcast. However, I'm fairly sure I used this with up to 5 or 6 locations while I was testing multiple weather locations to Zepp OS devices (since that is the limit imposed by the watch), but I do not recall what weather provider or locations I was using back then, they most likely less data.

This might be relevant to consider for #863 as well.

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open OR closed issue.
- [X] I checked that my issue is not covered in the [Frequently Asked Questions](https://github.com/breezy-weather/breezy-weather/blob/main/HELP.md) document.
- [X] I have updated the app to **[version 5.2.8 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)** and I can reproduce the issue in this version.
- [X] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance), unless my issue only happens with a specific language. I can write in my native language under the English text, if necessary.
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by papjul** [comment]
Indeed, it looks like it is limited to 1 MB.

So it's not the number of locations that matter but how much data they have. You could broadcast 20 locations with a source with only 5 days of forecast, while one with many details over 16 days such as Open-Meteo could be filled quickly. So limiting to 5 locations does not guarantee it will never crash.

What would you suggest instead?

**Comment by joserebelo** [comment]
Yeah, going by number of locations is not a solution - and I honestly do not have a good one, especially not one that would not be a breaking change.

1. We could limit the number of daily / hourly forecasts alongside the number of locations. It's still not fool-proof, but should allow us to compute an approximate maximum payload size per location.
2. We could add secondary locations up to a certain threshold, and stop then
3. We could switch from json to something else (parcelable, like the original WeatherSpec? it should take much less space than json)
4. A content provider or similar like I suggested in #863 would also remove this limitation

In the short term, adding some limits (points 1 and 2) might be the simplest path to mitigate this.

I can try and open a PR with some if you want.

**Comment by papjul** [comment]
The content provider looks like the cleanest solution from all of them, even though it will take some time (but it will also be useful for me). The first 3 are not fool-proof if there is a 1 MB limit at Android level. I may catch the error while we work on the alternative.

As it's not something that only affects Breezy Weather, I believe the content provider must be written in a way that works for all weather apps you would like to consume.

I'm not very familiar with that. Do you have some guidelines? Like the permission, I believe something like `nodomain.freeyourgadget.gadgetbridge.READ_WEATHER` (and maybe a separate `nodomain.freeyourgadget.gadgetbridge.READ_LOCATION_DETAILS`?)?

**Comment by papjul** [comment]
From what I can see, `ContentProvider` expects a `Cursor` which is a database-dependent thing. It's less flexible than JSON and its "Object" structure. Would it be possible to communicate through a `FileProvider` instead? I don't know if that's something recommendable for a read-only process.

**Comment by joserebelo** [comment]
> From what I can see, ContentProvider expects a Cursor which is a database-dependent thing. 

Yes, but not necessarily. You can easily return a MatrixCursor with whatever data you have. Take the simplest example from the one implemented in Catima, to return a single record with the major / minor version:

```java
final String[] columns = new String[]{Version.MAJOR_COLUMN, Version.MINOR_COLUMN};
final MatrixCursor matrixCursor = new MatrixCursor(columns);
matrixCursor.addRow(new Object[]{Version.MAJOR, Version.MINOR});

return matrixCursor;
```

Let's take the GadgetbridgeData object as an example, we would probably need URIs to query:

- list of locations
- current data for a location (by ID?) - single record
- hourly forecasts for a location (by location ID)
- daily forecasts for a location (by location ID)

The only thing that does not fit very well is the GadgetbridgeAirQuality object, which would most likely need the fields flattened on each of the above.

I think it would not be a bad API, but my vision is currently limited by the fields available in Gadgetbridge, I don't have full context on the full data that Breezy Weather can provide across all its providers.

> I'm not very familiar with that. Do you have some guidelines? Like the permission

I believe this inverts the API slightly. While with the WeatherSpec Gadgetbridge defined the API that it expects to receive, in this case it's Breezy Weather that defines it. With this, the permission would not be Gadgetbridge-specific, and any app could take advantage of the ContentProvider.

Similarly, I'm not sure it's expected for other weather apps to implement the same content provider interface, and if they wish to provide a similar one, Gadgetbridge would have to implement support for it.

Gadgetbridge could define an API contract (eg. Android defines one for [Calendar events](https://developer.android.com/reference/android/provider/CalendarContract), but that would be potentially limiting for the apps that implement it, especially if you want to provide a "Breezy Weather WeatherSpec" (now ContentProvider).

**Comment by joserebelo** [comment]
One small note about content providers: the authorities and permissions must be unique across apps, so you most likely want to declare them as `<breezy-weather-package>.something`, to avoid conflicts between the released app and debug app, should the user wish to have both installed in parallel (see https://github.com/CatimaLoyalty/Android/issues/1410)

**Comment by papjul** [comment]
I cannot guarantee location IDs will match between each query. I can only guarantee they will be unique (which is a point of an ID) amongst all locations sent at once.

What I can do is make 1 `ContentProvider` with multiple URI:
1. One for weather version just like you did for Catima to make the app aware of any compatiblity issue. Major = breaking change: renamed, changed type or deleted field. Minor = added field/feature
2. One for our own Json format
3. One for Gadgetbridge version, based on the version of the `WeatherSpec`
4. One for Gadgetbridge weather Json data

For the Gadgetbridge weather Json data URI, I can send a cursor with:
- One row per location
- Only one column (the json formatted)
- Each row contains the `GadgetbridgeData` json formatted

This will require you to have Breezy Weather-specific code in the Gadgetbridge app, but at least it will be more reliable and the object will be the same `GadgetbridgeData` you're used to, so the changes should be limited.

What do you think?


Other linked topic: I'm just not sure at this point how I can make `ContentObserver` work, so you can be notified of changes.

**Comment by joserebelo** [comment]
I think something as simple as the ID being the index of the location in Breezy weather should be enough stability (only potential race condition would be if some app queries 2 URIs right as the user is performing changes?).

> Only one column (the json formatted)

Searching online, I think that having a single json blob will pose the similar size issues, since there seems to be a maximum cursor row size: https://stackoverflow.com/questions/11162031/android-cursor-maximum-size

I would suggest at least splitting the URIs between locations / daily / hourly - assuming we do have IDs.

> 3. One for Gadgetbridge version, based on the version of the WeatherSpec
> 4. One for Gadgetbridge weather Json data

What is the difference between these?

> This will require you to have Breezy Weather-specific code in the Gadgetbridge app, but at least it will be more reliable and the object will be the same GadgetbridgeData you're used to, so the changes should be limited.

I believe this is fine. I also don't see a big problem in consuming Breezy Weather's own format, assuming that in the end it will provide the same data and be somewhat stable in the long term.

> Other linked topic: I'm just not sure at this point how I can make ContentObserver work, so you can be notified of changes.

I have never tried this, but looking online it should be as simple as something like this:

```java
getContext().getContentResolver().notifyChange(
    Uri.parse("content://org.breezyweather.provider/whatever"), 
    null);
```

**Comment by papjul** [comment]
> I think something as simple as the ID being the index of the location in Breezy weather should be enough stability - only potential race condition would be if some app queries 2 URIs right as the user is performing changes?

The way it was designed is that the whole database is rewritten when something changes, so if someone queries something from the database, it should be either everything or nothing for consistency IMO.

> Searching online, I think that having a single json blob will pose the similar size issues, since there seems to be a maximum cursor row size: https://stackoverflow.com/questions/11162031/android-cursor-maximum-size
> 
> I would suggest at least splitting the URIs between locations / daily / hourly - assuming we do manage to establish some sort of ID.

The issue happens when there are many locations. I don't think having one location per row should be an issue. At least in the way Gadgetbridge object is designed. For the Breezy Weather object, this will need some tests as there is more data.

It should also be more pleasant from consumer side to have a correctly formatted object and not just raw data.

>> 3. One for Gadgetbridge version, based on the version of the WeatherSpec
>> 4. One for Gadgetbridge weather Json data
> 
> What is the difference between these?

Number 3 sends the version "4" on one row. Number 4 sends the WeatherSpec with each row = one location. So that you can query the version first to ensure the compatibility.

**Comment by joserebelo** [comment]
Fair enough, I would just indeed keep the Breezy Weather object in mind.

> Number 3 sends the version "4" on one row. Number 4 sends the WeatherSpec with each row = one location. So that you can query the version first to ensure the compatibility.

I don't think it's worth the effort of maintaining both on the Breezy weather side, since Gadgetbridge currently does not support either and something will be implemented. Since there needs to be development anyway, it would also be fine to consume the Breezy weather one directly, reducing the maintenance effort on your side.
